platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDb.java

```diff
@@ -470,112 +485,33 @@ public <K extends VirtualKey, V extends VirtualValue> MerkleDbTableConfig<K, V>
     }
 
     /**
-     * Takes a snapshot of the database into the specified folder. Only primary open tables are
-     * included to snapshots.
+     * Takes a snapshot of the database source into the specified folder.
      *
      * @param destination Destination folder
      * @throws IOException If an I/O error occurred
      */
-    @SuppressWarnings("rawtypes")
-    public void snapshot(final Path destination) throws IOException {
```

Comment:
Do we ever need a snapshot of the entire state, rather than a single datasource?
Just wondering if we really want to only support single-datasource snapshots, or if we might need both options.

Reply:
This is the key problem this fix addresses. In the current implementation, a whole MerkleDb instance is snapshotted. It works fine at the database level, but unfortunately VM data is spread across two layers: virtual node cache and data source. There is no way to take a snapshot of all caches atomically, they are serialized one by one. It leads to inconsistency between node cache and data source.

Till this issue with virtual node cache is fixed, we have to snapshot data sources one by one, too.

Reply:
hmmm...  Sounds like something we'll need to resolve properly sometime soon, but perhaps not in this PR.